### PR TITLE
Implement a retry policy for deadlocks and other transient failures.

### PR DESCRIPTION
**Pull request by DimCitus**
After seeing the following happen in Travis, we could process some error
conditions as transient and retry. It would be best for test stability, and
also for users who script around some pg_autoctl commands.

    ERROR Monitor ERROR:  deadlock detected
    ERROR Monitor DETAIL:  Process 21381 waits for ShareLock on transaction 532; blocked by process 21382.
    ERROR Monitor Process 21382 waits for ExclusiveLock on advisory lock [16385,822708183,0,11]; blocked by process 21381.
    ERROR Monitor HINT:  See server log for query details.
    ERROR Monitor CONTEXT:  while updating tuple (0,11) in relation "node"
    ERROR Monitor SQL statement "UPDATE pgautofailover.node SET goalstate = $1, statechangetime = now() WHERE nodeid = $2"
    ERROR SQL query: SELECT pgautofailover.stop_maintenance($1)
    ERROR SQL params: '2'
    ERROR Failed to stop_maintenance of node 2 from the monitor
    FATAL Failed to stop maintenance from the monitor, see above for details

### CODE REVIEW

File: src/bin/pg_autoctl/monitor.c
Comment (JelteF) [inline src/bin/pg_autoctl/monitor.c:2609]: Can we use our exponential random backoff mechanism to implement this sleep?

Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c:2609]: Not easily. Our dynamic retry policy depends on a state that is attached to the PGSQL object, and used privately in the context of re-connecting. The policy we set on a monitor PGSQL object is meant to defer retries to another part of the code, and so we enforce `pgsql_set_main_loop_retry_policy` on the monitor's connection, which means no retry by default.

We could change the level of abstraction where our function `pgsql_compute_connection_retry_sleep_time` works and then handle that context in the monitor's function where we implement some retry, or maybe on their callers. I can have a try at that.

Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c]: Done now, including with the registration monitor API where we already had a crude retry mechanism.

File: src/bin/pg_autoctl/cli_enable_disable.c
Comment (JelteF) [inline src/bin/pg_autoctl/cli_enable_disable.c:445]: We use this same retry policy in 3 places. Can we give it a name that makes sense and make it a constant/function?

File: src/bin/pg_autoctl/pgsql.c
Comment (JelteF) [inline src/bin/pg_autoctl/pgsql.c:508]: Afaict this call does something completely different than the code that was here before.
1. It doesn't break when it reaches a signal, instead it enters this if block.
2. It does the opposite of what it previously did whit `true`/`false` for the `maxR`+`maxT`.

File: src/bin/pg_autoctl/monitor.c
Comment (JelteF) [inline src/bin/pg_autoctl/monitor.c:2615]: Let's do this everywhere. People expect `sleep` to be a function, even github syntax highlighting does that it seems.

```suggestion
			int sleepTime = pgsql_compute_connection_retry_sleep_time(retryPolicy);
```

File: src/bin/pg_autoctl/monitor.c
Comment (JelteF) [inline src/bin/pg_autoctl/monitor.c:2671]:
```suggestion
			log_warn("Failed to stop_maintenance of node %d on the monitor, "
```

File: src/bin/pg_autoctl/monitor.c
Comment (JelteF) [inline src/bin/pg_autoctl/monitor.c:2678]:
```suggestion
			return monitor_stop_maintenance(monitor, nodeId, retryPolicy);
```

File: src/bin/pg_autoctl/pgsql.c
Comment (JelteF) [inline src/bin/pg_autoctl/pgsql.c:306]: As also described in my other comment. I believe this should be `return true` and below `return false`.

**Comment by JelteF** [comment]
There's quite a lot of serious errors in this last changeset. Can you try to test it a bit better after fixing these? It's definitely possible that I missed some edgecases.

File: src/bin/pg_autoctl/cli_enable_disable.c
Comment (DimCitus) [reply src/bin/pg_autoctl/cli_enable_disable.c]: Good point. The abstraction to set retry policies was a level too high to easily do that, so I changed them to take a ConnectionRetryPolicy instead of a PGSQL argument, and added a new one.

File: src/bin/pg_autoctl/pgsql.c
Comment (DimCitus) [reply src/bin/pg_autoctl/pgsql.c:523]: I think your later comment is right, the return values have been inverted in that function.

File: src/bin/pg_autoctl/monitor.c
Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c]: Done now.

File: src/bin/pg_autoctl/monitor.c
Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c:2674]: Oops. Good catch. Fixed now.

File: src/bin/pg_autoctl/monitor.c
Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c]: Another good catch, thanks. Copy/paste and all.

File: src/bin/pg_autoctl/pgsql.c
Comment (DimCitus) [reply src/bin/pg_autoctl/pgsql.c]: Done that way, you're right.

**Comment by DimCitus** [comment]
> There's quite a lot of serious errors in this last changeset. Can you try to test it a bit better after fixing these? It's definitely possible that I missed some edgecases.

I just did a first pass over what you reported, will schedule more review time with QA and testing, and maybe a better approach to make sure this does the right thing. Turning to draft meanwhile.

**Comment by DimCitus** [comment]
See also #392 and #395, where we can't `C-c` out of the `pg_autoctl create` command when there's a loop over `monitor_register_node`. This PR should fix it with a proper retry mechanism that allows for signal handling.

**Comment by DimCitus** [comment]
I have now reviewed and finished this PR, it's again ready for review. The main thing is that we changed the registration process with proper transaction handling and this broke our “poor's man” retry implementation here. Good news, the new implementation is easier to review, makes more sense, and should be more robust in the wild.

**Comment by DimCitus** [comment]
Extra note: sometimes the Travis nightly build for the master branch fails, and that's because the `pg_autoctl enable maintenance` or `pg_autoctl disable maintenance` functions are hitting a deadlock on the monitor. This PR with the retry policy for those interactive commands should fix the flakiness of the nightly builds.

Of course sometimes we get this problem in branch/PR builds too, and then I usually restart the individual unlucky build.

File: src/bin/pg_autoctl/pgsql.c
Comment (JelteF) [reply src/bin/pg_autoctl/pgsql.c:530]: The first item still stands. With the new code it will log the connection error on any signal, with the old code it wouldn't. I'm not sure if we care about that though.

File: src/bin/pg_autoctl/keeper.c
Comment (JelteF) [inline src/bin/pg_autoctl/keeper.c:1250]:
```suggestion
	 * standby is concurently being added to the same group.
```

File: src/bin/pg_autoctl/monitor.c
Comment (JelteF) [inline src/bin/pg_autoctl/monitor.c:682]: Shouldn't this use `monitor_retryable_error` too?
```suggestion
		if (monitor_retryable_error(parseContext.sqlstate, STR_ERRCODE_OBJECT_IN_USE))
```

**Review by JelteF — CHANGES_REQUESTED** [review]

**Comment by JelteF** [comment]
I confirmed that this indeed solves the Ctrl+C issue that was reported in #392

File: src/bin/pg_autoctl/pgsql.c
Comment (DimCitus) [reply src/bin/pg_autoctl/pgsql.c:530]: Yeah I think it's better to log about “cancelling” the connection attempt, it might be what the user is trying to achieve with interrupting (C-c) the flow at this point.

File: src/bin/pg_autoctl/keeper.c
Comment (DimCitus) [reply src/bin/pg_autoctl/keeper.c]: Fixed, thanks!

File: src/bin/pg_autoctl/monitor.c
Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c]: Yeah good point, we can check also for the generic monitor_retryable_error, and add the specific error that we can also retry during registration.

**Review by JelteF — APPROVED** [review]
Blocked until this is merged https://github.com/citusdata/pg_auto_failover/pull/402, but otherwise good to merge.

**Pull request merged**
